includes/object-cache.php

```diff
@@ -2006,17 +2006,12 @@ protected function maybe_serialize( $data ) {
             return igbinary_serialize( $data );
         }
 
-        if ( is_array( $data ) || is_object( $data ) ) {
-            // phpcs:ignore WordPress.PHP.DiscouragedPHPFunctions.serialize_serialize
-            return serialize( $data );
-        }
-
         if ( $this->is_serialized( $data, false ) ) {
-            // phpcs:ignore WordPress.PHP.DiscouragedPHPFunctions.serialize_serialize
-            return serialize( $data );
+            return $data;
         }
-
-        return $data;
+        
+        // phpcs:ignore WordPress.PHP.DiscouragedPHPFunctions.serialize_serialize
+        return serialize( $data );
```

Comment:
Always serialize, except if it already was serialized? @naxvog thoughts?

Reply:
Does Redis currently contain double serialized data? If so, what were the use cases?

Reply:
```php
<?php
class Naxvog_Test {
        private $creation_time;
        private $serialization_time;
        public $some_prop;
        private $some_secret_prop;
        public function __construct() {
                $this->creation_time = time();
                $this->some_secret_prop = 'baz';
                $this->some_prop = 'bar';
        }
        public function __serialize() {
                return [
                        'serialization_time' => time(),
                        'creation_time' => $this->creation_time,
                ];
        }
        public function __unserialize( $data ) {
                $this->serialization_time = $data['serialization_time'];
                $this->creation_time = $data['creation_time'];
        }
}
add_action( 'init', function() {
        $a = new Naxvog_Test();
        $s = serialize( $a );
        #var_dump( $a, $s );
        #var_dump( unserialize( $s ) );
        if ( isset( $_GET['debug-set'] ) ) {
                wp_cache_set( 'naxvog_test_obj1', $a );
                wp_cache_set( 'naxvog_test_obj2', $s );
        }
        if ( isset( $_GET['debug-get'] ) ) {
                var_dump( wp_cache_get( 'naxvog_test_obj1' ) );
                var_dump( wp_cache_get( 'naxvog_test_obj2' ) );
        }
} );
```

Reply:
Should we check `is_serialized()` twice in `maybe_unserialize()`?

Reply:
I would rather prefer a conversion script running on plugin update resolving such issues but this might be a non trivial approach, a LUA script and will take some time (best suited for action scheduler).

Reply:
I can't think of any valid way to serialize twice. Sure there might be serialized data within a serialized object but this is expected behaviour.

Have to look again but I'm fairly sure that I have not found any double serialized key our docker dev environment. Will have a look on my production redis instance. Should be fairly easy to find.

Reply:
OK found some instances (ran vim searching for `[[:cntrl:]]s:[[:digit:]]\+:"O` on a copy of the rdb database):

- [Plugin Update Checker](https://github.com/YahnisElsts/plugin-update-checker) (option keys `puc_external_updates-[slug]`, `puc_external_updates_theme-[slug])
- WPML (option keys: `wpml_config_index` + others)
- [Easy Social Share Buttons](https://socialsharingplugin.com/) (seems to be using Plugin Update Checker as the option key `external_updates-easy-social-share-buttons` suggests)
- [Google Sitemap Generator](https://wordpress.org/plugins/google-sitemap-generator/) (option key `sm_status`)

Reply:
Can we simply solve this by running this twice?

https://github.com/rhubarbgroup/redis-cache/blob/fbfc86fc22b4df0c636b05e5ce5da0b548fd5b08/includes/object-cache.php#L1979-L1985

Reply:
Have to look in the code of those plugins to confirm that this is not intended. If not we should find the cause for this double serialization instead.

Reply:
Running `unserialize()` twice seems risky, in case plugins do their own serialization.